packages/generator/src/parser/edmx-parser.ts

```diff
@@ -1,111 +1,44 @@
 /* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
 
 import { PathLike, readFileSync } from 'fs';
-import path, { basename } from 'path';
-import { createLogger } from '@sap-cloud-sdk/util';
+import path from 'path';
 import { parse } from 'fast-xml-parser';
-import { forceArray } from '../generator-utils';
-import {
-  EdmxAssociation,
-  EdmxAssociationSet,
-  EdmxComplexType,
-  EdmxEntitySet,
-  EdmxEntityType,
-  EdmxFunctionImport,
-  EdmxMetadata
-} from './parser-types';
-
-const logger = createLogger({
-  package: 'generator',
-  messageContext: 'edmx-parser'
-});
-
-export function parseEdmxFromPath(edmxPath: PathLike): EdmxMetadata {
+import { EdmxMetadata as EdmxMetadataV2 } from './parser-types-v2';
+import { EdmxMetadata as EdmxMetadataV4 } from './parser-types-v4';
+import { parseBaseMetadata, getRoot } from './edmx-parser-common';
+import { parseEdmxV2 } from './edmx-parser-v2';
+import { parseEdmxV4 } from './edmx-parser-v4';
+
+export function parseEdmxFromPath(
+  edmxPath: PathLike
+): EdmxMetadataV2 | EdmxMetadataV4 {
   const edmxFile = readFileSync(path.resolve(edmxPath.toString()), {
     encoding: 'utf-8'
   });
   return parseEdmxFile(edmxFile, edmxPath);
 }
 
-function parseEdmxFile(edmx: string, edmxPath: PathLike): EdmxMetadata {
+function parseEdmxFile(
+  edmx: string,
+  edmxPath: PathLike
+): EdmxMetadataV2 | EdmxMetadataV4 {
   const parsedEdmx = parse(edmx, {
     ignoreAttributes: false,
     attributeNamePrefix: ''
   });
   const root = getRoot(parsedEdmx);
-
-  const metadata: EdmxMetadata = {
-    path: edmxPath,
-    fileName: basename(edmxPath.toString()).split('.')[0],
-    namespace: root.Namespace,
-    entitySets: parseEntitySets(root),
-    entityTypes: parseEntityTypes(root),
-    functionImports: parseFunctionImports(root),
-    complexTypes: parseComplexTypes(root),
-    associationSets: parseAssociationSets(root),
-    associations: parseAssociations(root),
-    selfLink: parseLink(root)
+  const metaData = parseBaseMetadata(
+    root,
+    getODataVersion(parsedEdmx),
+    edmxPath
+  );
+
+  return {
+    ...metaData,
+    ...(metaData.oDataVersion === 'v2' ? parseEdmxV2(root) : parseEdmxV4(root))
```

Comment:
try to find all swtiches between v2 and v4 and use the isV2() method you introduced already everywhere so that it is easier to find all the places where a switch is.